### ISSUE TITLE
Bug: `no-restricted-imports` doesn't distinguish between package names that have a substring as a name

### ISSUE THREAD

**Issue by bengry** [eslint/eslint#18126]
### Environment

Node version: 20.1.0
npm version: N/A
Local ESLint version: 8.56.0
Global ESLint version: 8.56.0
Operating System: macOS 14.3.1

### What parser are you using?

@typescript-eslint/parser

### What did you do?

```js
/*eslint no-restricted-imports: ["error", {
        "patterns": [
          {
            "group": [
              "react"
            ],
            "importNames": [
              "forwardRef"
            ],
            "message": "Use our type-safe wrapper from \"foo\" instead."
          }
        ]
      }]*/

import { forwardRef as originalForwardRef } from "react";
import { forwardRef as modifiedForwardRef } from "@my-org/utils/react";

```

### What did you expect to happen?

should error:
```js
import { forwardRef as originalForwardRef } from 'react' 
```

shouldn't error:
```js
import { forwardRef as modifiedForwardRef } from '@my-org/utils/react'
```

### What actually happened?

Both `import`s error.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7XG4gICAgICAgIFwicGF0dGVybnNcIjogW1xuICAgICAgICAgIHtcbiAgICAgICAgICAgIFwiZ3JvdXBcIjogW1xuICAgICAgICAgICAgICBcInJlYWN0XCJcbiAgICAgICAgICAgIF0sXG4gICAgICAgICAgICBcImltcG9ydE5hbWVzXCI6IFtcbiAgICAgICAgICAgICAgXCJmb3J3YXJkUmVmXCJcbiAgICAgICAgICAgIF0sXG4gICAgICAgICAgICBcIm1lc3NhZ2VcIjogXCJVc2Ugb3VyIHR5cGUtc2FmZSB3cmFwcGVyIGZyb20gXFxcImZvb1xcXCIgaW5zdGVhZC5cIlxuICAgICAgICAgIH1cbiAgICAgICAgXVxuICAgICAgfV0qL1xuXG5pbXBvcnQgeyBmb3J3YXJkUmVmIGFzIG9yaWdpbmFsRm9yd2FyZFJlZiB9IGZyb20gJ3JlYWN0J1xuaW1wb3J0IHsgZm9yd2FyZFJlZiBhcyBtb2RpZmllZEZvcndhcmRSZWYgfSBmcm9tICdAbXktb3JnL3V0aWxzL3JlYWN0J1xuIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIn19fQ==

### Additional comments

I tried changing the rule configuration slightly, but to no avail.

Note that using this:
```json
{
  "name": "react",
  "importNames": ["forwardRef"],
  "message": "..."
}
```
Is not good enough since it now errors on all of these lines:
```js
import * as React from 'react';
```
Which, while technically correct (since you can then use `React.forwardRef`) is too much of an overhead, and not what I want to fix here. I'll prevent usage of `React.forwardRef` using another rule (`no-restricted-syntax maybe`).

**Comment by bengry** [comment]
> Note that this also disallows `import * as React from 'react'`.

Thanks. This is something I wanted to avoid:

> Note that using this:
> ```json
> {
>   "name": "react",
>   "importNames": ["forwardRef"],
>   "message": "..."
> }
> ```
> Is not good enough since it now errors on all of these lines:
> ```js
> import * as React from 'react';
> ```
> Which, while technically correct (since you can then use `React.forwardRef`) is too much of an overhead, and not what I want to fix here. I'll prevent usage of `React.forwardRef` using another rule (`no-restricted-syntax maybe`).

Is there any way to achieve what I want? Seems like a missing use-case for this rule. Perhaps an `exact: true` option could be added to the object?

**Comment by mdjermanovic** [comment]
> Is there any way to achieve what I want?

Currently not, `import *` is always disallowed. You can open a new rule change request.
